### PR TITLE
feat: limit number of files during storage cleanup

### PR DESCRIPTION
**Pull request by crepererum**
Since the number of parquet files can potentially be unbound (aka very
very large) and we do not want to hold the transaction lock for too
long and also want to limit memory consumption of the cleanup routine,
let's limit the number of files that we collect for cleanup.

### CODE REVIEW

File: parquet_file/src/cleanup.rs [hunk 1]
```diff
@@ -72,31 +76,30 @@ where
...
-                        .map(|part| part.encoded().ends_with(".parquet"))
-                        .unwrap_or(false)
-                        && !all_known.contains(&path_parsed)
-                })
-                .collect::<Vec<_>>()
-        })
-        .try_concat()
-        .await
-        .context(ReadError)?;
+    let mut to_remove = vec![];
```
Comment (alamb) [inline parquet_file/src/cleanup.rs:79]: (i was confused at first about what lock was held, but then I see that it is the [`transaction`](https://github.com/influxdata/influxdb_iox/pull/1621/files#diff-e79131d9debf7e45b83937dbfd0554def726eb989efff20bdd1dcdddb59aaf7cL48) 👍  that is preventing new files from being written while this loop is running.

File: server/src/db.rs [hunk 2]
```diff
@@ -965,7 +965,7 @@ impl Db {
                             debug!(?duration, "cleanup worker sleeps");
                             tokio::time::sleep(duration).await;
 
-                            if let Err(e) = cleanup_unreferenced_parquet_files(&self.catalog).await {
+                            if let Err(e) = cleanup_unreferenced_parquet_files(&self.catalog, 1_000).await {
```
Comment (alamb) [inline server/src/db.rs:968]: I don't have any sense of how long it takes to delete 1000 files  but my intuition is that it is on the order of seconds. I wonder if we should crank down the number another order of magnitude (100?)

**Review by alamb — COMMENTED** [review]
I think the code does what it was designed to and is better than `main`

As written this code still holds an open catalog transaction while listing from object store which might take a long time and is not under our direct control.   I wonder if we could invert the logic to minimize the time the transaction was held open:

1. Record a snapshot of all object store files that currently exist
2. open the transaction and get the list of all known files, and drop the transaction
3. Remove all object store files that are not in the known list

This sequence of operations might miss some files that were both added and removed from the catalog between the snapshot of the object store and the catalog inventory, but they would be removed in the next round

However, this PR seems better than master so 👍  to merge it from me. If we do cap the number of files that are deleted per iteration I think we may also need to increase the frequency of the background loop. Since we only run the cleanup loop every 500 seconds and it will delete at most 1000 files each iteration, the cleaner won't be able to keep up if more than 2 parquet files per second are created on average.

**Review by alamb — APPROVED** [review]

**Comment by crepererum** [comment]
@alamb see #1623 aka I think that won't work w/o race conditions at the moment.

File: server/src/db.rs (hunk 2 above)
Comment (crepererum) [reply server/src/db.rs:968]: The deletion itself does not hold the transaction lock, so that should be fine.

**Comment by crepererum** [comment]
> If we do cap the number of files that are deleted per iteration I think we may also need to increase the frequency of the background loop. Since we only run the cleanup loop every 500 seconds and it will delete at most 1000 files each iteration, the cleaner won't be able to keep up if more than 2 parquet files per second are created on average.

Let's keep an eye on it and decide later how we wanna tune these parameters. At least this PR should make the node less crashy (although I still think that the deadlock we where seeing was another bug, since holding the transaction lock should not lockdown the whole DB because "write chunk to OS" is a background task).

**Comment by alamb** [comment]
> Let's keep an eye on it and decide later how we wanna tune these parameters. 

I think this sounds like a good plan

**Pull request merged**